### ISSUE TITLE
Minetest has telemetry that can't be disabled / missing privacy consent

### ISSUE THREAD

**Issue by hecktest** [minetest/minetest#14819]
### Minetest version

```true
5.x.something, this is pretty recent
```

### Summary

MT now phones home to check for content updates. Countless issues:

- This change wasn't very widely announced for how heavy handed it is. I found out about it by noticing an error message when no network connection was available. How many versions has it been pinging this server for?
- The network request happens on startup, previously MT only made a network request when manually and intentionally clicking the "join game" tab.
- **There is no option to disable this check.** This option should exist and it should be disabled by default.
- The code performing the function is barely documented and it'll take me some precious time to find out what exactly is being sent and received.
- The code refers to checking for updates for installed content. ~~Since there are way too many packages in existence to download the full list, it probably sends the server the installed package list.~~
~~- Does the code narrow it down to packages that came from CDB or do you literally receive a list with "mibi" in it every time I connect?~~
- Even without private package names, chances are that a user's package set and device configuration is unique enough to form a fingerprint.  This, with the IP, might be enough to reconstruct a list of the places I've been staying at for the past few months. Why do you want this data, @rubenwardy ? What does it realistically help you with?
- Do you literally do this just to display a f****ng number and some icons in the formspec? This feels like the HTTP request itself is the point, and the GUI just an excuse.
- The code seems confined to the content form, why is it running on startup anyway?
- Even if the CDB server source is available, it is impossible to verify that this exact code is running on the server, and that data isn't being illegally retained. Once it's sent, it's sent.
- This is a huge leap from the previous situation of only receiving a list from a master server, without any data being *sent*.
- You seem to have a privacy policy on the site; you have to go out of your way to find it at the bottom and there are multiple ways to receive this software without having a chance to encounter its text. 
- Not only that, the policy is blatantly lying, saying you have to click the CDB tab for network requests to occur, which is no longer true. This, combined with the above, is enough to get you in legal trouble.
- "Platform, OS, device, etc" sounds very broad and vague, not to mention 100% unnecessary for a repository of mods made in a **platform-independent** language such as Lua.
- Several free software operating systems keep Minetest in their software repositories, even very strict, security conscious ones like OpenBSD. I think most of these maintainers have a mental image of MT matching the reality from a few versions ago, and are blisfully unaware of this sharp turn, just like I was.

Overall, this is behavior I'd expect from mobile trash, not from free software.

Tagging @rubenwardy because this is his pet project. Very, very disappointed in this direction. I should have been more vigilant when this feature was first being championed because _of course_ it was going to lead to this, people who want to build shiny online systems for everything will eventually shove updates, telemetry, analytics and other trash down your throat.

People have hard forked for less. I'll kick MT off my firewall whitelist for the time being; I'll wait and see how you handle this issue before making any move.

### Steps to reproduce

Download the shitware called Minetest, enable terminal logging and try to run it without a network connection. Notice a new message in the terminal.

**Comment by rubenwardy** [comment]
This is very important: **Minetest only checks for updates if you have at least one package installed using the ContentDB menu integration and it never sends the package list to ContentDB.**

> The code refers to checking for updates for installed content. Since there are way too many packages in existence to download the full list, it probably sends the server the installed package list.

It never sends the package list - it's a static endpoint, only the Minetest version (and standard request stuff) is sent. Ie: https://content.minetest.net/api/updates/?protocol_version=43&engine_version=5.8.0

> There is no option to disable this check. This option should exist and it should be disabled by default.

I'm not against an option but is should be enabled by default as it provides the best user experience

> The code performing the function is barely documented and it'll take me some precious time to find out what exactly is being sent and received.

It's 13 lines of Lua code: https://github.com/minetest/minetest/blob/master/builtin/mainmenu/content/update_detector.lua#L50-L63

> The network request happens on startup, previously MT only made a network request when manually and intentionally clicking the "join game" tab.

This is not true, there's also the version check to www.minetest.net

> Does the code narrow it down to packages that came from CDB or do you literally receive a list with "mibi" in it every time I connect?

No package lists are sent to ContentDB

**Comment by sfan5** [comment]
The update check was added in #7629 (5.6.0), the CDB update check in #13807 (5.8.0).

> Platform, OS, device, etc" sounds very broad and vague, not to mention 100% unnecessary for a repository of mods made in a platform-independent language such as Lua.

For HTTP requests it makes Minetest assembles an user agent with its own versions as well as the version of your OS.
for example:
* `"Minetest/5.8.0 (Windows/10.0.19041 x86_64)"`
* `"Minetest/5.9.0-dev-fcb4f258f (Linux/6.9.7-arch1-1 x86_64)"`

This is not technically necessary for any of our use cases. But it has helped us with abuse prevention and gauging which versions are still in use and where in the past.
Personally I don't think this is unreasonable tracking for a FOSS app, *if* the user consents.

> This change wasn't very widely announced for how heavy handed it is.

We don't have a special process for announcing changes that touch privacy. It's a boring change log entry like everything else.
Would be good to have one, since this is also of interest to distros/downstreams.

> I think most of these maintainers have a mental image of MT matching the reality from a few versions ago, and are blisfully unaware of this sharp turn, just like I was. [...] Speaking of conscience, here's an example of a repo that you've duped with this change:

This is obviously on us. But an argument could be had on how much due diligence downstreams are expected to perform, when they didn't notice an anti-feature being added. Think "trust but verify".

FWIW Debian is aware and has [disabled the update check](https://salsa.debian.org/games-team/minetest/-/blob/master/debian/rules?ref_type=heads#L28). I don't see any changes related to CDB in their patches.

<hr>

In short:
* Our privacy policy is incorrect. This is a problem.
* Are we required to get informed consent from the user before making HTTP requests? However as a good natured FOSS app we should do this anyway.
  * this should be implemented by having a dialog with a short text and a link to the privacy policy
* Downstreams are often not aware of privacy-impeding features and how to turn them off at build-time.
* In addition to that users should have the choice to turn off the CDB package check at startup

**Comment by sfan5** [comment]
> The fact that something as mainstream as Ubuntu has a policy like this should raise a flag that maybe this is something you shouldn't be doing, not a "haha, let's do this, maybe they won't notice, dumb users don't know what's good for them"

The context of this statement is that Debian (and with it Ubuntu) is notoriously slow to update which causes worse user experience, wasted support effort for ancient versions and in part also leads to a situation where modders have to consider supporting old versions, because those are in wider use than they should be.
You can grab Debian 12 or Ubuntu 22.04 fresh of the presses right now and receive a finely aged (almost) 2 year old Minetest version. (We stopped recommending them [on the website](https://github.com/minetest/minetest.github.io/pull/296) for this reason.)
It is in Debian/Ubuntu's interest to disable update checks as to not confuse users because they consider their package manager to be the authoritative source of updates, which is where the conflict lies.

I say this to be clear that "We want to add tracking which Ubuntu rightfully prohibits but maybe they won't notice" is absolutely not the intent of what ruben said.

**Comment by kromka-chleba** [comment]
>I don't know if a repo should ping for updates like this, no decent software repository I know of does this unprompted.

I second this, honestly just an "update package list" would do. This is a minor inconvenience, yet provides privacy by default. Also for users who don't care that much you can add "enable automatic package list updates" or something.

>Either it passes their standard for phoning home (incorrectly imo, since minetest now needs packages to function), or they're unaware because it's not easily discovered.

MTG was so bad, most players would jump to CDB for mods/games anyway.

>Right now, my personal safety could depend on your server's security, and I didn't even know about this until yesterday.

Although I agree with you that Minetest should ask before doing **any** connections, I think that having such high levels of trust for a ... **computer game** especially with such heavy security concerns is wrong in the first place. Your security/life shouldn't depend on a game and if it does then maybe unplug the machine from the internet.


Anyway, just 5 cents from someone who doesn't know what he's talking about.

**Comment by NathanSalapat** [comment]
Just to throw in my two cents, github and literally every website are collecting far more information about you than Minetest, and yet you seem to have no trouble using them.

An option/setting to disable update checking may be warranted, but freaking out like this isn't. :D

**Comment by kromka-chleba** [comment]
@hecktest Okay, I understand your point more now and agree this is serious.
Though you still shouldn't trust random software by default, especially when you have a commercial project that could get compromised because of it. Unless you can explain to me why having trust is good for security.

Providing a flag to disable CDB integration and also making `ENABLE_UPDATE_CHECKER=FALSE` the default would be a good idea.

**Comment by kromka-chleba** [comment]
Some distros could want to disable CDB integration to encourage using packages from the repository?
Guix for example has packages for minetest games and mods that way. (though CDB is still there by default).

**Comment by AFCMS** [comment]
IMO, would be cool to have the option to disable the check, maybe with a popup at first startup asking something like "Enable online services?", but honestly checking for package updates is a great feature overall.

I personally tend to clone Git repositories for my mods/games and auto update everything regularly (I use CDB to discover them anyways), but for non-technical users or mobile phones having a notification for available updates is an important feature. Without it I would often forget to check for update on the content tab before starting to play on Android.

I strongly disagree that this feature should be off by default.

And really, why would sending a stupid MT version + OS as user agent be a problem?? What could CDB do with this data? A graph with used MT versions and OS at best.  Since it's how internet works, CDB will have access to my IP (which reveal my rough location) unless I use a VPN, this is much more impact-full data in case the CDB owner becomes malicious. I am for passing the Minetest version, as it's useful to estimate how much are the different versions used. OS isn't really needed I guess but I honestly don't care.

> This is also a bad practice, especially now that any commit is a donation of code for MS Copilot to launder.

This project is open-source. As long as you have a public repo there is nothing you can do if someone wants to gather code to feed an AI model.

Yes, GitHub train an AI. So does GitLab and Bitbucket. But even if you host your own instance of Forgejo, Gitea or GitLab, you have plenty of web crawlers solutions that are able to scan the entire internet, look for instances of these services and download the code. And you can't even block at all if they use a random user agent. Any publicly available data can be gathered by any people or robots.

AI being ethical or not is another debate, but complaining about your code being used by GitHub while building a open-source software which is by definition accessible by anyone is ridiculous.

**Comment by HybridDog** [comment]
I'm against automatic update checks when Minetest starts unless I have explicitly enabled them.

> If this was up to me, I'd forbid having any home-pings on by default in every distribution except these two cases:
> […]

Does NTP time synchronisation also count as a home-ping?

**Comment by lhofhansl** [comment]
I agree with @hecktest .

We should have a simple rule for Minetest: "No network requests, unless explicitly requested by the user."
Anything else, IMHO, will not do.

(And I do not care what Github, or the OS, or whatever are doing, this is about Minetest.)

**Comment by jordan4ibanez** [comment]
I honestly think we should have a debian mode which will allow 2 stages of removal of all these pertaining issues:

Something along the lines of:

```c++
enum Telemetry {
  // Will allow the game to work as it does now, unchanged.
  full,
  // Will disable the engine from sending out any os information but can access server list and contentdb.
  slight, 
  // Will disable the engine from talking to the server list and contentdb, with the contentdb tab removed.
  none
};
```
In this implementation, the lua code will detect that the engine flag, somehow. Subsequently executing only the desired implementation.

With ``full`` it would just work as it currently does.

Having ``slight`` it will be able to talk to the server list and contentdb, but it will refuse to give out system information with placeholders instead.

If ``none`` is enabled, there should be a hard code in the engine which if it detects a tab is created on the main menu with the name ``contentdb`` it could simply close the game. It could also detect if it's trying to send out to the contentdb url or server list, completely closing the game as well.

This is simply an idea of how to make everyone happy depending on how the end user and linux distro maintainers want to package or compile the game. This has been a long time issue so I thought I would finally chime in with my full thoughts after processing the distress around this issue.

Let me know your thoughts on it.

**Comment by SPiRiT369** [comment]
Just thought it's worth to mention:
1. Setting `update_information_url` to blank should disable checking for new minetest version.
2. You can control the contentdb url with `contentdb_url`.
3. You can set a fake url to fetch servers list with `serverlist_url`.

https://github.com/minetest/minetest/blob/master/builtin/settingtypes.txt
